addons/mail/static/src/messaging/utils/throttle/throttle.js

```diff
@@ -0,0 +1,426 @@
+        }
+        const coolingDownDeferred = makeDeferred();
+        this._coolingDownDeferred = coolingDownDeferred;
+        this._isCoolingDown = true;
+        const cooldownTimeoutId = this.env.window.setTimeout(
+            () => coolingDownDeferred.resolve(),
+            this._duration
+        );
+        let unexpectedError;
+        try {
+            await coolingDownDeferred;
+        } catch (error) {
+            if (
+                !(error instanceof _ThrottleFlushedError) ||
+                error.throttleId !== this.id
+            ) {
+                // This branching should never happens.
+                // Still defined in case of programming error.
+                unexpectedError = error;
+            }
+        } finally {
+            this.env.window.clearTimeout(cooldownTimeoutId);
+            this._coolingDownDeferred = undefined;
+            this._isCoolingDown = false;
+        }
+        if (unexpectedError) {
+            throw unexpectedError;
+        }
+    }
+
+}
+
+//------------------------------------------------------------------------------
+// Public
+//------------------------------------------------------------------------------
+
+/**
+ * A function that creates a cancelable, flushable and clearable throttle
+ * version of a provided function. See definitions at the top of this file.
+ *
+ * This throttle mechanism allows calling a function at most once during a
+ * certain period:
+ *
+ * - When a function call is made, it enters a 'cooldown' phase, in which any
+ *     attempt to call the function is buffered until the cooldown phase ends.
+ * - At most 1 function call can be buffered during the cooldown phase, and the
+ *     latest one in this phase will be considered at its end.
+ * - When a cooldown phase ends, any buffered function call will be performed
+ *     and another cooldown phase will follow up.
+ *
+ * @param {Object} env the OWL env
+ * @param {function} func the function to throttle.
+ * @param {integer} duration duration, in milliseconds, of the cooling down
+ *   phase of the throttling.
+ * @param {Object} [param2={}]
+ * @param {boolean} [param2.interceptCancelationErrors=false] if set, caller
+ *   of throttle function will intercept some errors that come from current
+ *   throttle call that has been canceled, such as when throttle function has
+ *   been explicitly canceled with `.cancel()` or when another new throttle call
+ *   has been registered.
+ *   @see ThrottleCanceledError for when a call has been canceled from explicit
+ *     call.
+ *   @see ThrottleReinvokedError for when a call has been canceled from another
+ *     new throttle call has been registered.
+ * @returns {function} the cancelable, flushable and clearable throttle version
+ *   of the provided function.
+ */
+function throttle(
+    env,
+    func,
+    duration,
+    { interceptCancelationErrors = false } = {}
```

Comment:
By design these errors should propagate to business code, because it could have an `await` on these timers and manage cancelation in a certain way. At the moment it's operational to keep promise indefinitely pending, because we don't need to do special stuffs when such errors happen.

The annoyance of propagating these errors in business code is to require `try {} catch {}`/`.catch()` whenever we invoke a throttle function, which makes some noise...

...But at the same time, what if we have another kind of error in a nested asynchronous function (programming error, network error, etc...)? I think we are not used to error management code in Odoo, which certainly explain why I see it as noise at the moment.

We could not support this and always require `try {} catch {}`/`.then()`. What do you think?

Reply:
Until we need it in business code it's dead code, and we usually don't want to keep dead code...
I'm fine if we keep it for now because this is a technical detail that shouldn't delay/prevent merging this feature, but if we really never use it at the end, I'm not sure if we should keep the option. We can always reintroduce it when needed.

Reply:
As discussed, try to "always intercept", which leads to removing the "errors", which leads to also getting rid of deferred.